gh-84116: Docs: Document help and aliases for argparse.add_parser() (#140574)
python/cpython: Doc/library/argparse.rst

Reviewer: I would actually recommend adding the formal documentation for the `~_SubParsersAction.add_parser` method just above this block. We can remove this redundant section about aliases and incorporate `deprecated` as well.

I'm thinking something like:

```
.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)
<rest of method docs>

This returned :class:`!ArgumentParser` object can be modified as usual.

The following example shows how to use *aliases* to create a shorthand
for a subcommand, like ``svn`` uses ``co`` for ``checkout``::
<examples>
```

Author: Hi @savannahostrowski ,
i have made the changes and also the changes in the free threading doc is done in another PR #140566 which is merged.
please let me know if you need any further changes

Reviewer: I'd recommend adding the new docs in here and then reconciling the `alias`/`deprecated` docs below into your content.

Author: Added the new doc and removed the previous discussion about the aliases and deprecated.

Reviewer 2: Please add back this documentation! We only want to document `_SubParserAction.add_parser` but we don't want to change other documentation.

Author: I have added the changes only having `_SubParserAction.add_parser`.

Reviewer 2: According to the diff, no. I don't see where the `required` parameter is documented. Please first fix the conflicts.

Reviewer 2: For instance where is this line? it's not here anymore.

Author: added.

Other utilities
---------------

> Subcommands

Reviewer 2: I think we can revert this change. Changing it would change the generated indices and thus break existing URLs, which is not always a good idea.

Author: This changes was mentioned by @savannahostrowski ,Based on the #142488 in previous discussion.

Reviewer: As I mentioned before, this was hugely inconsistent in the docs. I would prefer if this was not reverted.

Author: Hi @savannahostrowski ,
Actually i have done the changes for `Sub-commands` to `Subcommands` but @picnixz said to revert as it was before as it break the docs consistency. I guess that's what he meant, If the final change is to change the   `Sub-commands` to `Subcommands` then i will revert it.

Reviewer 2: I actually thought that the changes were merged into main but it appears not. You can keep what you wrote.

Author: Thanks @picnixz  for your review,
I reverted the change but in @savannahostrowski  pr https://github.com/python/cpython/pull/142488  she mentioned about the `Sub-commands` so i clearly reverted it. but will we also need to shift the `sub-parser` to `subparser`.

Author: Reverted the changes from `Sub-commands` to `Subcommands` as suggested in the pr #142488 by @savannahostrowski

   :class:`ArgumentParser` supports the creation of such subcommands with the
   :meth:`!add_subparsers` method.  The :meth:`!add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object
   has a single method, :meth:`~_SubParsersAction.add_parser`:

   .. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

      Creates and returns a new :class:`!ArgumentParser` object for the
      subcommand *name*.

      The *name* argument is the name of the subcommand.

>       The *help* argument provides a short description for this subcommand. If provided, it will be listed next to the command in the main parser's help message (e.g., ``PROG --help``).

Reviewer 2: Please wrap lines under 80 characters.

Author: The changes are done. Sorry for my inconvenience.

      The *aliases* argument allows to provide a sequence of strings that can be used as alternative names for this subcommand (e.g., ``aliases=['r']`` for a ``'run'`` command).

      The *deprecated* argument, if :const:`True`, marks the subcommand as deprecated, which typically issues a warning when used.

>       All other keyword arguments are passed directly to the

Reviewer 2: ```suggestion

      All other keyword arguments are passed directly to the
```

      :class:`!ArgumentParser` constructor.

>       This returned :class:`!ArgumentParser` object can be modified as usual.

Reviewer 2: ```suggestion
```

   Description of parameters:

>    * *title* - title for the subparser group in help output; by default

Reviewer 2: Revert those unrelated changes. Or more generally, why was this changed?

Author: This is mentioned by @savannahostrowski  in previous discussion. But as it causes link breaking i have changed it back to its original form.

   :class:`ArgumentParser` supports the creation of such subcommands with the
   :meth:`!add_subparsers` method.  The :meth:`!add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object

>    has a single method, :meth:`~_SubParsersAction.add_parser`:

Reviewer 3: ```suggestion
   has a single method:
```
I suggest de-duplicating, you are linking to text directly below (you can find more on this in the devguide).

   :class:`ArgumentParser` supports the creation of such subcommands with the
   :meth:`!add_subparsers` method.  The :meth:`!add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object
   has a single method, :meth:`~_SubParsersAction.add_parser`:

.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

   Creates and returns a new :class:`!ArgumentParser` object for the
   subcommand *name*.

   The *name* argument is the name of the sub-command.

   The *help* argument provides a short description for this sub-command.
   If provided, it will be listed next to the command in the main parser’s
   help message (e.g., ``PROG --help``).

   The *aliases* argument allows providing alternative names for this
   sub-command.

   The *deprecated* argument, if ``True``, marks the sub-command as
   deprecated and will issue a warning when used.

   All other keyword arguments are passed directly to the
   :class:`!ArgumentParser` constructor.

>    Examples

Reviewer 2: Don't add a new section. Just add the examples close to the parameters they refer to. The usage examples should, however, be kept after the description of the method.

Author: The new section is removed and they are placed in there respective area.

   :meth:`!add_subparsers` method.  The :meth:`!add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object
   has a single method, :meth:`~_SubParsersAction.add_parser`:

.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

   Creates and returns a new :class:`!ArgumentParser` object for the
   subcommand *name*.

   The *name* argument is the name of the sub-command.

   The *help* argument provides a short description for this sub-command.
   If provided, it will be listed next to the command in the main parser’s
   help message (e.g., ``PROG --help``).

   The *aliases* argument allows providing alternative names for this
   sub-command.

   .. versionadded:: 3.14

   For example::

      >>> parser = argparse.ArgumentParser()
      >>> subparsers = parser.add_subparsers()
      >>> checkout = subparsers.add_parser('checkout', aliases=['co'])
      >>> checkout.add_argument('foo')
      >>> parser.parse_args(['co', 'bar'])
      Namespace(foo='bar')

   The *deprecated* argument, if ``True``, marks the sub-command as
   deprecated and will issue a warning when used.

>    .. versionadded:: 3.14

Reviewer 2: This is wrong. It was added in 3.13.

Author: Rectified it.

   :class:`ArgumentParser` supports the creation of such subcommands with the
   :meth:`!add_subparsers` method.  The :meth:`!add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object
   has a single method, :meth:`~_SubParsersAction.add_parser`:

   .. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

      Creates and returns a new :class:`!ArgumentParser` object for the
      subcommand *name*.

>       The *name* argument is the name of the sub-command.

Reviewer: ```suggestion
      The *name* argument is the name of the subcommand.
```

Reviewer: We mostly use "subcommand" vs. "sub-command", trying to get us to a consistent state 🙂 . See https://github.com/python/cpython/pull/142488

>       The *help* argument provides a short description for this sub-command. If provided, it will be listed next to the command in the main parser's help message (e.g., ``PROG --help``).

Reviewer: ```suggestion
      The *help* argument provides a short description for this subcommand. If provided, it will be listed next to the command in the main parser's help message (e.g., ``PROG --help``).
```

Author: Thank you  for the correction, i have made the changes accordingly.

>       The *aliases* argument allows to provide a sequence of strings that can be used as alternative names for this sub-command (e.g., ``aliases=['r']`` for a ``'run'`` command).

Reviewer: ```suggestion
      The *aliases* argument allows you to provide a sequence of strings that can be used as alternative names for this subcommand (e.g., ``aliases=['r']`` for a ``'run'`` command).
```

Author: Thank you  for the correction, i have made the changes accordingly.

   :class:`ArgumentParser` supports the creation of such subcommands with the
   :meth:`!add_subparsers` method.  The :meth:`!add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object
   has a single method, :meth:`~_SubParsersAction.add_parser`:

   .. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

      Creates and returns a new :class:`!ArgumentParser` object for the
      subcommand *name*.

>       :param name: The name of the sub-command.

Reviewer 2: Please avoid using `:param:` roles & co. Explain each documented parameter in its own paragraph as for the rest of the file.

Author: Hi @picnixz,
I have made the requested changes; please review again. Let me know if you need any change.

   :class:`ArgumentParser` supports the creation of such subcommands with the
   :meth:`!add_subparsers` method.  The :meth:`!add_subparsers` method is normally
   called with no arguments and returns a special action object.  This object
   has a single method, :meth:`~_SubParsersAction.add_parser`, which takes a
   command name and any :class:`!ArgumentParser` constructor arguments, and
   returns an :class:`!ArgumentParser` object that can be modified as usual.
   has a single method, :meth:`~_SubParsersAction.add_parser`:

   Description of parameters:
.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

>    * *title* - title for the sub-parser group in help output; by default

Reviewer 2: The section "descriptions of parameters" is tied to `ArgumentParser.add_subparsers`. Yoiu are documenting `_SubParsersAction.add_parser` which is different. In other words, don't change the "Description of parameters" section.

Author: Thank you @picnixz ,
Okay i will undo the changes, but will i add the block for `_SubparsersAction.add_parser`  just below the example part of the `Argument.add_subparsers`

   Description of parameters:

>    * *title* - title for the sub-parser group in help output; by default "subcommands" if description is provided, otherwise uses title for positional arguments

Reviewer 2: Ok I don't why you're trying to change this. Why are changing this part of the document? I feel my reviews are just being ignored.

Author: Added the changes, check it out .Sorry @picnixz  if you feel that way but it is not what actually happening. Please check it and review it.

   Description of parameters:

   * *title* - title for the sub-parser group in help output; by default
      "subcommands" if description is provided, otherwise uses title for

>       positional arguments

Reviewer 2: I'm not sure about those whitespaces. This should be aligned with the parameter itself as it's currently done.

Author: Hi @picnixz , 
Here doing only t his indentation change may break the whole description of parameter indentation if you want i can fix the indentation issue of this like i can change every other parameter description in this format.

Reviewer 2: Can you show us a screenshot please? because the current indentation is correct though?

Author: <img width="650" height="570" alt="image" src="https://github.com/user-attachments/assets/12926809-5e4b-4a3f-850c-2f01037b64fa" />
Hello @picnixz,
Here is my description of parameter block from my side. Tell me if i need to do anything.

Reviewer 2: I wanted a screenshot of the docs after they are built. Revert those whitespaces and check the generated docs. I believe the current indentation is correct. It is wrong to have this indent.

Author: <img width="1049" height="803" alt="image" src="https://github.com/user-attachments/assets/75a214a8-7ae2-4ba3-a8e1-b8e10c7d3cf8" />

Hello @picnixz ,
here is what my side of the generated doc look like.

Reviewer 2: This one is with your changes? because currently we don't have this if you look online:

<img width="1739" height="938" alt="image" src="https://github.com/user-attachments/assets/4e094770-431b-4ff4-ad02-2eaeca167255" />

So please, make sure that your generated docs don't have this additional indent.

Author: Thanks for the rectification @picnixz ,
I have changed it and this what my documentation look like. Please review it.
<img width="1128" height="555" alt="image" src="https://github.com/user-attachments/assets/3ab92526-487e-44c0-9470-51089c2543fa" />

   for that particular parser will be printed.  The help message will not
   include parent parser or sibling parser messages.  (A help message for each
   subparser command, however, can be given by supplying the ``help=`` argument

>    to ``add_parser()`` as above.)

Reviewer 2: ```suggestion
   to :meth:`~_SubParsersAction.add_parser` as above.)
```

This text is far enough from the previous mention to `add_parser` to warrant another link.

Author: changed according to your request.

       {foo,bar}   additional help

   Furthermore, :meth:`~_SubParsersAction.add_parser` supports an additional

>    *aliases* argument,

Reviewer 2: Be sure that the versionadded is included.

Author: The versionadded are also included.

       {foo,bar}   additional help

>    .. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

Reviewer 2: Why do we have a duplicated entry here?

      Creates and returns a new :class:`!ArgumentParser` object for the
      subcommand name.

      The name argument is the name of the subcommand.

>       The help argument provides a short description for this subcommand.

Reviewer 2: Merge the documentation but chose whether to place it here or elsewhere. It doesn't make sense to have duplicated text. As for the examples, either put them below the documented method or in a dedicated section.

Author: Done the changes. Please tell me if there is any further issues

      the main parser.

.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None,

>                                            deprecated=False, **kwargs)

Reviewer 2: You need to aligned deprecated with name.

Author: Done.

   Creates and returns a new :class:`!ArgumentParser` object for the
   subcommand *name*.

   The *name* argument is the name of the sub-command.

   The *help* argument provides a short description for this sub-command.
   If provided, it will be listed next to the command in the main parser’s
   help message (for example, ``PROG --help``).

   The *aliases* argument allows providing alternative names for this
   sub-command.

   For example::

      >>> parser = argparse.ArgumentParser()
      >>> subparsers = parser.add_subparsers()
      >>> checkout = subparsers.add_parser('checkout', aliases=['co'])
      >>> checkout.add_argument('foo')
      >>> parser.parse_args(['co', 'bar'])
      Namespace(foo='bar')

   The *deprecated* argument, if ``True``, marks the sub-command as
   deprecated and will issue a warning when used.

   For example::

      >>> parser = argparse.ArgumentParser(prog='chicken.py')
      >>> subparsers = parser.add_subparsers()
      >>> fly = subparsers.add_parser('fly', deprecated=True)
      >>> parser.parse_args(['fly'])  # doctest: +SKIP
      chicken.py: warning: command 'fly' is deprecated
      Namespace()

   .. versionadded:: 3.13

>       Added the *deprecated* parameter.

Reviewer 2: This was meant to be added after "constructor.". And be careful with the number of blank lines. Don't just ask an LLM to make them. In fact, don't use LLMs for formatting.

Author: @picnixz ,Please just give me the direction i am not able to get your point about adding the `.. versionadded` in the deprecated part.

      Subparser's *prog* is no longer affected by a custom usage message in
      the main parser.

> .. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)

Reviewer 2: ```suggestion

.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None, deprecated=False, **kwargs)
```

And wrap the signature under 80 characters.

Author: changed it please review it.

   Creates and returns a new :class:`!ArgumentParser` object for the
   subcommand *name*.

   The *name* argument is the name of the sub-command.

   The *help* argument provides a short description for this sub-command.
   If provided, it will be listed next to the command in the main parser’s

>    help message (e.g., ``PROG --help``).

Reviewer 2: ```suggestion
   help message (for example, ``PROG --help``).
```

Author: Changed according to your request.

   The *aliases* argument allows providing alternative names for this
   sub-command.

   For example::

      >>> parser = argparse.ArgumentParser()
      >>> subparsers = parser.add_subparsers()
      >>> checkout = subparsers.add_parser('checkout', aliases=['co'])
      >>> checkout.add_argument('foo')
      >>> parser.parse_args(['co', 'bar'])
      Namespace(foo='bar')

   The *deprecated* argument, if ``True``, marks the sub-command as
   deprecated and will issue a warning when used.

>    .. versionadded:: 3.13

Reviewer 2: Now that I read the text, I think we should move this versionadded at the end of the method description with `Added the *deprecated* parameter`

Author: Added the changes.

   For example::

      >>> parser = argparse.ArgumentParser(prog='chicken.py')
      >>> subparsers = parser.add_subparsers()
      >>> fly = subparsers.add_parser('fly', deprecated=True)
      >>> parser.parse_args(['fly'])  # doctest: +SKIP
      chicken.py: warning: command 'fly' is deprecated
      Namespace()

> 
>    All other keyword arguments are passed directly to the
>    :class:`!ArgumentParser` constructor.

Reviewer 2: ```suggestion

   All other keyword arguments are passed directly to the
   :class:`ArgumentParser` constructor.

```

Author: changed according your request.

      the main parser.

> .. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None,

Reviewer 2: ```suggestion
.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None,\
```

When the signature requires multi lines I think we need this additional break. But if the docs render well without it you do not need to add it.

Author: i think it is better to not add the \ as it disrupts the indentation

                                         deprecated=False, **kwargs)

   Create and return a new :class:`ArgumentParser` object for the
   subcommand *name*.

   The *name* argument is the name of the sub-command.

   The *help* argument provides a short description for this sub-command.

   The *aliases* argument allows providing alternative names for this
   sub-command. For example::

      >>> parser = argparse.ArgumentParser()
      >>> subparsers = parser.add_subparsers()
      >>> checkout = subparsers.add_parser('checkout', aliases=['co'])
      >>> checkout.add_argument('foo')
      >>> parser.parse_args(['co', 'bar'])
      Namespace(foo='bar')

   The *deprecated* argument, if ``True``, marks the sub-command as
   deprecated and will issue a warning when used. For example::

      >>> parser = argparse.ArgumentParser(prog='chicken.py')
      >>> subparsers = parser.add_subparsers()
      >>> fly = subparsers.add_parser('fly', deprecated=True)

>       >>> args = parser.parse_args(['fly'])  # doctest: +SKIP

Reviewer 2: Can you remove the doctest skip? and write down the exact output so that we also catch message regressions just in case? ty

Author: Hi @picnixz , Even after removing the `# doctest: +SKIP` the document does not produce the warning. so if you want i can remove the warning line below.

Reviewer 2: If the warning line is not emitted that is a bit weird. Is the warning emitted with the REPL?

Author: @picnixz i am able to produce the warning now 
the warning goes by 
`chicken.py: warning: command 'fly' is deprecated`
`Namespace()`
so i will be adding exactly the same or not.

Reviewer 2: This has still not be changed. I want you to remove the skip and match the message correctly..

Author: I was just updating my branch here i have added the changes you requested.

      the main parser.

> .. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None,

Reviewer 4: Missing backslash, fixed at https://github.com/python/cpython/pull/144572.

      the main parser.

.. method:: _SubParsersAction.add_parser(name, *, help=None, aliases=None,
                                         deprecated=False, **kwargs)

>    .. versionadded:: 3.13

Reviewer 2: I want this at the end of the section for that method. You put it at the begjnning. That is sftr the "All other ..."

Author: Thanks a lot for your feedback. I made the changes which you have said.

>    Creates and returns a new :class:`!ArgumentParser` object for the

Reviewer 2: ```suggestion
   Create and return a new :class:`ArgumentParser` object for the
```

   subcommand *name*.

   The *name* argument is the name of the sub-command.

   The *help* argument provides a short description for this sub-command.
   If provided, it will be listed next to the command in the main parser’s

>    help message (for example, ``PROG --help``).

Reviewer 2: ```suggestion
```

I think this part is trivially evident and is not worth documenting.

   The *aliases* argument allows providing alternative names for this
   sub-command.

>    For example::

Reviewer 2: ```suggestion
   The *aliases* argument allows providing alternative names for this
   sub-command. For example::
```

      >>> parser = argparse.ArgumentParser()
      >>> subparsers = parser.add_subparsers()
      >>> checkout = subparsers.add_parser('checkout', aliases=['co'])
      >>> checkout.add_argument('foo')
      >>> parser.parse_args(['co', 'bar'])
      Namespace(foo='bar')

   The *deprecated* argument, if ``True``, marks the sub-command as
   deprecated and will issue a warning when used.

>    For example::

Reviewer 2: ```suggestion
   The *deprecated* argument, if ``True``, marks the sub-command as
   deprecated and will issue a warning when used. For example::
```

      >>> parser = argparse.ArgumentParser(prog='chicken.py')
      >>> subparsers = parser.add_subparsers()
      >>> fly = subparsers.add_parser('fly', deprecated=True)
      >>> parser.parse_args(['fly'])  # doctest: +SKIP
      chicken.py: warning: command 'fly' is deprecated

>       Namespace()

Reviewer 2: ```suggestion
      >>> args = parser.parse_args(['fly'])  # doctest: +SKIP
      chicken.py: warning: command 'fly' is deprecated
```

      >>> parser.parse_args(['fly'])  # doctest: +SKIP
      chicken.py: warning: command 'fly' is deprecated
      Namespace()

   All other keyword arguments are passed directly to the

>    :class:`!ArgumentParser` constructor.
> 

Reviewer 2: ```suggestion
   :class:`!ArgumentParser` constructor.

```

   Sometimes it's desirable to use a custom string in error messages to provide
   more user-friendly output. In these cases, :meth:`!register` can be used to

>    register custom actions or types with a parser allows you to reference the

Reviewer 2: This is wrong. Revert.

Author: The mistake was reverted.

   Sometimes it's desirable to use a custom string in error messages to provide
   more user-friendly output. In these cases, :meth:`!register` can be used to

>    register custom actions or types with a parser anto reference the

Reviewer 2: Why was this changed?

Author: Sorry, my mistake . It happened unintentionally.
